# Profitable App Profiles for the App Store and Google Play Markets

## Introduction

For this project, I am acting as a data analyst working for a company that builds free Android and iOS apps. As this means the company's main source of revenue is in-app ads, the number of app users determines the total revenue for any given app. My goal here is to analyze data to help developers understand what type(s) of app is likely to attract more users.

## Getting Started: Opening and Exploring the Data

I am going to focus on datasets for apps on Google Play and the App Store, as that is my company's market. First, I'm going to open two datasets, one for each platform.

In [1]:
google_opened_file = open('googleplaystore.csv')
from csv import reader
google_read_file = reader(google_opened_file)
google_apps_data = list(google_read_file)

In [2]:
ios_opened_file = open('AppleStore.csv')
from csv import reader
ios_read_file = reader(ios_opened_file)
ios_apps_data = list(ios_read_file)

I opened each file, making sure to distinguish through the variable names which dataset they were assigned to. I imported the `reader()` function and used that to read in the files, and from that, transformed each read-in file to a list of lists. Now that the files are in this form, I will be able to explore them.

Below I will do so using the `explore_data()` function and print the first three rows.

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

This function takes in four parameters:

- `dataset`, which will be a list of lists
- `start` and `end`, which will both be integers and represent the starting and ending indices of a slice from the dataset
- `rows_and_columns`, which will be a Boolean and has `False` as a default argument

And it slices the dataset using `dataset[start:end]` and prints the number of rows and columns if `rows_and_columns` is `True`. One important thing to note is `dataset` should not have a header row.

In [4]:
print(google_apps_data[0])
print('\n')
explore_data(google_apps_data, 1, 4, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10842
Number of columns: 13


In [5]:
print(ios_apps_data[0])
print('\n')
explore_data(ios_apps_data, 1, 4, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7198
Number of columns: 16


Most of these column names are adequately descriptive, but some may not be. [Here is a link](https://www.kaggle.com/datasets/lava18/google-play-store-apps) to the dataset documentation for the Google Play app dataset, and [here is a link](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps) for the iOS dataset. The compilers include descriptions for each column name.

For our purposes, the columns that may be of interest to us in the Google Play dataset are `'App'`, `'Category'`, `'Reviews'`, `'Installs'`, `'Type'`, `'Price'`, and `'Genres'`.

For the App Store dataset, they are `'track_name'`, `'price'`, `'rating_count_tot'`, and `'prime_genre'`.

## Cleaning the Data

Now that we've opened and done some preliminary exploring of our datasets, we might be excited to do some analysis!

Not so fast. I need to clean the data first. This entails detecting inaccurate data, so I can correct or remove it, as well as detecting and deleting duplicates.

I am keeping in mind that at this company, we only build apps that are *free to download* and for an *English-speaking* audience. This means I also need to remove both non-English apps and paid apps from the dataset.

### Deleting Wrong Data

Let's recall the dataset documentation links. Along with the Data Card and Code sections, there is also a Discussion section. I'll go to the Google Play link first. One user was kind enough to point out that entry 10472 has a wrong value in the `'Rating'` column. Let's check that out by printing the row in question.

In [6]:
print(google_apps_data[0])
print(google_apps_data[10473])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


We do have a problem here. The `'Rating'` value is indeed wonky but this is actually because the row is missing a value in the `'Category'` column, causing the entire row to shift. Let's just delete this row from our dataset.

In [7]:
del google_apps_data[10473]

It's important to remember that the above `del` statement should not be run more than once, or else I'll end up deleting extra rows. We don't want that.

Let's see which row ended up taking our deleted row's place.

In [8]:
print(google_apps_data[10473])
print('\n')
print("New dataset length:", len(google_apps_data))

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


New dataset length: 10841


Looking good. Notice the dataset now has one fewer row. Upon further browsing of the Google Play dataset discussion, another user alerted us to the issue of duplicates. I'm going to paste (a lightly edited version of) what he wrote:

> ...it's not proper to judge duplicated apps just by their names. For example, there are two apps from different types (game and family) both named **Solitaire**, but you will lose this information if you only judge duplication by apps' name.

This sounds pretty important to know. He then added:

> ...for a certain app named **Cardiac diagnosis (heart rate, arrhythmia)**, it has both a free version and paid version, if you do the same thing like you did on **Solitaire**, you will miss the free data or the paid data which will influence your analysis on all free apps.

Excellent. I'll be keeping this in mind as I go through more data cleaning. This seems to be it for the Google Play discussion.

Now for the App Store dataset discussion. As of writing, I have only come across one report of wrong data. One user mentioned that there are two duplicates in this dataset but did not specify which ones. Looks like I'll have to pick them out myself. All this talk of duplicates here brings us to our next part...

### Removing Duplicate Entries

Based on the documentation discussions, it seems pretty apparent that the datasets, especially the Google Play one, have duplicate entries. Let's confirm this by printing some duplicate rows:

In [9]:
for app in google_apps_data:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Instagram alone has four entries!

Before I start deleting willy-nilly, I need to find a way to first identify all duplicates and then figure out, of the multiple entries, which to keep and which to delete.

The best way to accomplish the first part is to create two lists, one for duplicate apps and another for unique apps. Then I would start looping through all the apps (entries) in a given dataset and, using the `list.append()` method, sort each app into one of the two lists.

For the Google Play apps, it would look like this:

In [10]:
duplicate_apps_google = []
unique_apps_google = []

for app in google_apps_data[1:]:
    name = app[0]
    if name in unique_apps_google:
        duplicate_apps_google.append(name)
    else:
        unique_apps_google.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps_google))
print('Expected dataset length:', len(google_apps_data[1:]) - len(duplicate_apps_google))
print('\n')
print('Examples of duplicate apps:', duplicate_apps_google[:15])

Number of duplicate apps: 1181
Expected dataset length: 9659


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


That (mostly) takes care of my first task. I want to briefly address the points made earlier in the Google Play discussion about duplicates. After trying several times to write code that accounts for "false duplicates" and failing, I have concluded two things:

1. I do not have the skills and knowledge necessary to write code that can account for several conditions (columns in this case), even though it has been established that looping through app name alone produces faulty (albeit trivially so) results.
2. A better way to complete this task, within my knowledge scope, would be to include another column in the master list of lists with the app's store URL. Assigning each app this piece of information, or some other type of unique ID, should solve this issue.

For the second task, I need to develop a criterion for removing duplicates rather than deleting randomly.

Looking back at the multiple Instagram entries, the only difference between the rows is the number of reviews (index `3`). This indicates the data was collected at different times. That is, the higher the number of reviews, the more recent the data should be. So, I'll keep the row with the highest number of reviews and remove all others.

Now it's time to actually get to removing the duplicates. To do this, I'm going to create a dictionary where each key is a unique app name and the corresponding value is the highest number of reviews for that app.

In [11]:
reviews_max = {}
for row in google_apps_data[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(reviews_max['Instagram'])
print('Expected dictionary length:', len(google_apps_data[1:]) - len(duplicate_apps_google))
print('Actual dictionary length:', len(reviews_max))

66577446.0
Expected dictionary length: 9659
Actual dictionary length: 9659


So far so good. For the next step, I'm going to use the information stored in the dictionary and create a new dataset, which will have only one entry (the one with the highest reviews number) per app.

In [12]:
android_clean = []
already_added = []

for row in google_apps_data[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(row)
        already_added.append(name)
        
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


What exactly did I just do here?

1. As mentioned earlier, I created a dictionary called `reviews_max` to then use for removing duplicates.
2. I then initialized two empty lists, `android_clean` and `already_added`.
    1. `android_clean` stores the new cleaned data.
    2. `already_added` stores only app names. I'll explain the need for this list in a minute.
3. I looped through the Google Play dataset minus the header row (`google_apps_data[1:]`). For each iteration, I:
    1. Assigned the app name to a variable called `name`.
    2. Converted the number of reviews to `float`, then assigned it to a variable called `n_reviews`.
    3. Added the current app (`row`) to the `android_clean` list and the app name (`name`) to the `already_added` list if:
        1. The number of reviews for the current app (`n_reviews`) matches the number of app reviews stored in the `reviews_max` dictionary, **and**
        2. The name of the app is not in the `already_added` list. This supplementary condition is necessary to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If I only check for `reviews_max[name] == n_reviews`, I'll still end up with duplicate entries for some apps.
4. Outside of the loop, I used the `explore_data` function for the first three rows to ensure everything went as expected, and set the last parameter to `True` to confirm the amount of rows and columns.

What about the iOS App Store?

Upon further inspection of the dataset, it appears that the comment in the discussion was a false alarm. There are no duplicates I can see, and in this case, I was able to check using the `id` instead of the `track_name` column. So we can safely move onto the next step of data cleaning.

### Removing Non-English Apps

Unlike with the duplicates, both datasets here do have apps with names that suggest they are not designed for an English-speaking audience. Here's an example.

In [13]:
print(ios_apps_data[814][1])
print(ios_apps_data[6732][1])
print('\n')
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ


One way to filter out some of these apps is to remove any app name containing a symbol that isn't commonly used in English text. This would include letters in the English alphabet, numbers composed of digits 0-9, punctuation marks (., !, ?, ;) and other symbols (+, *, /).

Each character used in a string has a corresponding number associated with it. For instance, the corresponding number for character `a` is 97, character `A` is 65, and character `爱` is 29,233. The corresponding number for any character can be found using the `ord()` built-in function.

In [14]:
print(ord('a'))
print(ord('A'))
print(ord('爱'))
print(ord('5'))
print(ord('+'))

97
65
29233
53
43


Conveniently for us, all characters commonly used in English text are in the range 0 to 127, according to the [ASCII](https://en.wikipedia.org/wiki/ASCII) (American Standard Code for Information Interchange) system. Based on this range, I can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character is used in English text. Likewise, if an app name contains a character that is greater than 127, then it probably means the app has a non-English name.

Based on this, I can write a function that takes a string as a parameter and returns `False` if any character in the string doesn't belong to the set of common English characters. Otherwise, the function will return `True`.

In [15]:
def is_english(app_name):
    for character in app_name:
        if ord(character) > 127:
            return False
    return True

To test this function, I'm going to input some various app names as arguments:

In [16]:
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


Looks like this test exposed some holes in my function. The last two apps tested are both in English, but the function failed to identify them as so. This is because emojis and characters like `™` fall outside the ASCII range and have corresponding numbers over 127.

In [17]:
print(ord('™'))
print(ord('😜'))

8482
128540


If I use the above function as is, I'll lose useful data since many English apps will be incorrectly labeled as non-English. To rectify this, I'll set a different threshold and only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This filter function is still not perfect, but it should be fairly effective.

In [18]:
def is_english(app_name):
    non_ascii = 0
    for character in app_name:
        if ord(character) > 127:
            non_ascii += 1
            if non_ascii > 3:
                return False
    return True

And to test the new function:

In [19]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Great. Using this new function, I'm going to loop through each dataset. If an app name is identified as English, I'm going to append the whole row to a new list. Starting with the Google Play dataset:

In [20]:
google_english_apps = []

for row in android_clean:
    app_name = row[0]
    if is_english(app_name) == True:
        google_english_apps.append(row)
        
explore_data(google_english_apps, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


So far so good. Onto the iOS apps:

In [21]:
ios_english_apps = []

for row in ios_apps_data[1:]:
    app_name = row[1]
    if is_english(app_name) == True:
        ios_english_apps.append(row)
        
explore_data(ios_english_apps, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


It appears we have **9,614** Android apps and **6,183** iOS apps left after removing the non-English apps.

### Isolating the Free Apps

The last part of data cleaning will be to isolate all free apps. Remember that the company only builds apps that are free to download and install, and our main source of revenue consists of in-app ads.

For this, I'm going to loop through each dataset and put all the free apps in separate lists. Since prices will be coming up as strings, I'll have to make sure I'm not checking an `integer` or a `float` in my conditional statements.

In [22]:
# As usual I'll start with Android apps

android_apps_final = []

for row in google_english_apps:
    price = row[7]
    if price == '0':
        android_apps_final.append(row)
        
print(len(android_apps_final))

8864


In [23]:
ios_apps_final = []

for row in ios_english_apps:
    price = row[4]
    if price == '0.0':
        ios_apps_final.append(row)
        
print(len(ios_apps_final))

3222


It looks like I'll ultimately be working with **8,864** Android apps and **3,222** iOS apps. Now I'm finally ready to move onto analyzing the data.

## Analyzing the Data

### Most Common Apps by Genre: Part One

As I mentioned in the introduction, this project's goal is to determine the kinds of apps that are likely to attract more users because of the number of people using my (LARP) company's apps affect our revenue.

To minimize risks and overhead, the company's validation strategy for an app idea has three steps:

1. Build a minimal Android version of the app and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

I'm going to begin the analysis by determining the most common genres for each market. For this, I'll need to build frequency tables for a few columns in the datasets. Let's take a look at the columns again and see which ones can lend themselves best to the frequency tables.

In [24]:
# Android apps

print(google_apps_data[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


`'App'`, `'Category'`, and `'Genres'` stand out. `'Rating'` and `'Reviews'` can also be useful to gauge popularity and user response.

In [25]:
# iOS apps

print(ios_apps_data[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


`'track_name'` and `'prime_genre'` stand out. `'rating_count_tot'` and `'user_rating'`, as well as `'rating_count_ver'` and `'user_rating_ver'` can also be useful to gauge popularity and user response.

### Part Two

I'm going to build two functions to analyze the frequency tables:
- One function to generate frequency tables that show percentages
- Another function to display the percentages in a descending order

As we know, the frequency tables will be built in the dictionary format. However, dictionaries don't have order, so how will I analyze the frequency tables? This question will explain the need for the second function that will display the entries in the frequency table in descending order.

To do that, I will make use of the built-in `sorted()` function. It takes in an iterable data type, like a list, dictionary, tuple, etc., and returns a list of the elements of that iterable sorted in ascending or descending order.

The `sorted()` function doesn't work very well with dictionaries because it only considers and returns the dictionary keys. However, if I transform the dictionary into a list of tuples, where each tuple contains a key along with its corresponding value, the sorting can work right.

This is a bit overcomplicated to just sort a dictionary, but I hear there are much simpler ways to do this once I learn more advanced techniques. For now though, I'm going to create a helper function called `display_table()` which does make use of this workaround. It will do the following:

- Take in two parameters, `dataset` and `index`. `dataset` will be a list of lists, and `index` will be an integer.
- Generate a frequency table using the `freq_table()` function, which I will go into more detail on soon.
- Transform the frequency table into a list of tuples, then sort the list in a descending order (hence the `reverse = True` argument)
- Print the entries of the frequency table in descending order

First though, let's get that prerequisite `freq_table()` function out of the way.

In [26]:
def freq_table(dataset, index):
    
    new_table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in new_table:
            new_table[value] += 1
        else:
            new_table[value] = 1
            
    table_percentages = {}
    for key in new_table:
        percentage = (new_table[key] / total) * 100
        table_percentages[key] = percentage
        
    return table_percentages

The `freq_table()` function takes in two inputs, `dataset` (a list of lists) and `index` (an integer). It is a garden variety frequency table creator, but because I want the frequencies expressed as percentages, I made a few modifications.

For instance, I included a variable `total` that tallies up the number of entries. When the function finishes running, `total` should have the same value as `len(dataset)`.

After the initial frequency table, I initialized another dictionary called `table_percentages` that only expresses percentages instead of absolute values. Within its for loop, it makes use of `total` as well in its percentage calculations.

Now back to the `display_table()` function.

In [27]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Let's see what frequency tables we get for the Android app columns `'Genres'` and `'Category'`, and then for the iOS app column `'prime_genre'`.

In [28]:
# 'Genres'

display_table(android_apps_final, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

In [29]:
# 'Category'

display_table(android_apps_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [30]:
# 'prime_genre'

display_table(ios_apps_final, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


### Part Three

Let's take a closer look at what the information in the frequency tables tells us.

I'm going to start with the `'Category'` and `'Genres'` columns from the Google Play dataset. For `'Category'`, "Family" is the largest group by far, with 18.9% of apps. Second and third place are "Game" and "Tools", with 9.7% and 8.5% respectively. The rest of the groups all contain shares of under 5%. For `'Genres'`, "Tools" is the largest group with about 8.5% of apps. Second and third place are "Entertainment" and "Education", with 6.1% and 5.3% respectively. These are the only groups that contain more than 5% of apps; the rest fall under. This column is also more granular, and for now we are more interested in the big picture that `'Category'` can provide. The distribution for `'Genres'` also seems slightly more uniform than that of `'Category'`, which is a bit more right skewed.

Generally, in these two columns, the types of apps that tend to dominate are tools, games (most of "family" likely belongs here), entertainment, lifestyle, productivity, and business/finance. A good mix of apps for both practical and entertainment purposes, but among these two tables, practical seems to win out just in terms of app numbers.

As for the `'prime_genre'` column of the App Store dataset, this distribution is extremely right skewed, with "Games" taking up a whopping 58.2% of apps. "Entertainment" and "Photo & Video" are a distant second and third, comprising 7.9% and 5% of apps respectively. All other groups are well below 5%, with "Education" afterwards at 3.7%.

This distribution is very unlike what we just saw with the Google Play apps. The majority of App Store apps are games or serve other entertainment purposes.

Based on the App Store frequency table alone, I am inclined to recommend a certain app profile (gaming and entertainment) for this particular market, but it comes of course with caveats. One is that a large number of apps in a particular genre does not necessarily imply that apps of that genre will have a large number of users. However, this is likely not a problem for iOS users. Another is that while a gaming app sounds "tried and true," the field is crowded. Although a truly successful gaming app can bring in mountains of revenue, it is just as likely an app here can get lost in the crowd.

For Google Play, practical apps fill a much larger share in comparison to entertainment apps/games. With that said, the same caveats I mentioned for the App Store will probably apply here as well. The frequency tables I generated reveal the most frequent app genres, not the genres that have the most users.

### Most Popular Apps by Genre on the App Store

To get a better picture of which app genres truly are more popular, instead of relying on common app genres, I'm going to shift my focus to calculating the average number of installs for each app genre. For the Google Play dataset, this information is in the `'Installs'` column, but the App Store dataset has no such column. As a workaround, I'll take the total number of user ratings as a proxy, which is represented in the `'rating_count_tot'` column.

To calculate the average, I will take the following steps:
1. Isolate the apps of each genre
2. Add up the user ratings/installs for the apps of that genre
3. Divide the sum by the number of apps belonging to that genre

In [31]:
ios_genres = freq_table(ios_apps_final, 11)

for genre in ios_genres:
    total = 0
    len_genre = 0
    for row in ios_apps_final:
        genre_app = row[11]
        if genre_app == genre:
            n_ratings = float(row[5])
            total += n_ratings
            len_genre += 1
    avg_ratings = total / len_genre
    print(genre, ":", avg_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


According to this table, the genres with high average amounts of user ratings are "Navigation" (highest) followed by "Reference" and "Social Networking." This information seems surprising after my analysis of the most common apps by genre. Let's look into "Navigation" a little further:

In [32]:
for row in ios_apps_final:
    if row[11] == 'Navigation':
        print(row[1], ':', row[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


No wonder we found "Navigation" to be surprising - Waze and Google Maps account for the vast majority of ratings and have cornered the market here. Let's look into "Reference" next:

In [33]:
for row in ios_apps_final:
    if row[11] == 'Reference':
        print(row[1], ':', row[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


In this case, the Bible and the dictionary similarly account for the vast majority of user ratings. It's now apparent that using average user ratings alone to gauge genre popularity is not the most reliable metric due to skewed distributions.

(As an aside, remember when I was filtering out non-English apps and remarked my method wasn't entirely perfect? We can see an example of that here in the penultimate entry. Thankfully the effect is almost negligible against the rest of the dataset.)

I have repeated the same for loop for "Social Networking", "Music", and other genres, and generally the same phenomenon occurs in which a few giants inflate average rating numbers and make the entire genre seem more popular than it actually is. Other genres are out of the question, like "Food & Drink." This one mostly consists of restaurant apps and delivery services, which falls well outside our scope. Another, like "Weather," may get a lot of eyes, but users do not generally spend a lot of time on weather apps.

For a concrete app profile recommendation, I am hesitant to suggest an entertainment or game app. The market seems saturated already. However, there seems to be more potential in a practical app. Reference and travel in particular come to mind.

For instance, we could develop an app that serves as a directory for a certain type of business or public service, like Gasbuddy, but instead for something like public restrooms or dog parks. The directory can have enhanced features like maps, directions, suggestions, and user ratings and reviews.

Another idea is to take a popular book like the Bible and create an app in which readers can interact with the text as well as access additional features (like illustrations and quizzes) to enrich their reading experience and understanding.

### Most Popular Apps by Genre on Google Play

The Google Play dataset does have an `'Installs'` column, but upon further inspection, it does not give precise numbers, only open-ended thresholds (100+, 1,000+, 5,000+, etc.). We can see for ourselves here:

In [34]:
display_table(android_apps_final, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


Since I'm working with big picture metrics, this isn't horrible news. I'm going to get started preparing these numbers to perform computations. This firstly entails converting each install number from a string to a float, which includes removing all commas and plus characters. This also means that in my analysis, I will be considering an app with 100,000+ installs to just have 100,000 installs. Again, not perfect, but it does the job.

To remove characters from strings, I'm going to use the `str_replace(old, new)` method. It takes in two parameters, `old` and `new`, and replaces all occurences of `old` within a string with `new`.

As with the iOS apps, I will start by generating a frequency table for the `'Category'` column of the Google Play dataset, then loop over the unique genres.

In [35]:
android_categories = freq_table(android_apps_final, 1)

for category in android_categories:
    total = 0
    len_category = 0
    for row in android_apps_final:
        category_app = row[1]
        if category_app == category:
            n_installs = row[5]
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            total += float(n_installs)
            len_category += 1
    avg_installs = total / len_category
    print(category, ":", avg_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

Due to the formatting of the numbers, this list isn't the easiest to read, but I'm going to highlight that the "Communication" category has the highest amount of installs, with more than 38 million. Will this be another extremely skewed distribution?

In [36]:
for row in android_apps_final:
    if row[1] == 'COMMUNICATION' and (row[5] == '1,000,000,000+'
                                      or row[5] == '500,000,000+'
                                      or row[5] == '100,000,000+'):
        print(row[0], ':', row[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

Yes. WhatsApp, Facebook Messenger, Skype, and Google Chrome each have more than a billion installs, with a good number of others surpassing 100 million. For the love of hypotheticals, what if I reexamined the "Communications" category after disregarding these popular apps?

In [37]:
under_100_m = []

for row in android_apps_final:
    n_installs = row[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (row[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
print(sum(under_100_m) / len(under_100_m))

3603485.3884615386


The average would be reduced by about ten times!

This pattern is present in a number of other popular categories like "Video Players" (second highest), "Social", "Photography", and "Productivity". As with the iOS apps, the averages make the categories appear more popular than they actually are, and the giants would be hard to compete with.

I am going to stay away from games right now for reasons mentioned earlier, and will instead focus my attention on a category that showed potential for success on the App Store, since the company's aim is to create something profitable for both markets. Let's look at "Books and Reference".

In [38]:
for row in android_apps_final:
    if row[1] == 'BOOKS_AND_REFERENCE':
        print(row[0], ':', row[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

There looks to be a healthy variety of apps here: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. Let's see if this category is highly skewed like some of the others.

In [39]:
for row in android_apps_final:
    if row[1] == 'BOOKS_AND_REFERENCE' and (row[5] == '1,000,000,000+'
                                            or row[5] == '500,000,000+'
                                            or row[5] == '100,000,000+'):
        print(row[0], ':', row[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


Although the category still has a few small giants, there are enough moderately successful apps, which could present us with a viable opportunity. Let's look at some of these apps to get ideas. I'm going to exclude any app with fewer than 1 million and more than 50 million installs.

In [40]:
for row in android_apps_final:
    if row[1] == 'BOOKS_AND_REFERENCE' and (row[5] == '1,000,000+'
                                            or row[5] == '5,000,000+'
                                            or row[5] == '10,000,000+'
                                            or row[5] == '50,000,000+'):
        print(row[0], ':', row[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there will be some significant competition.

We also notice there are quite a few apps built around the Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions

In this project, I analyzed data about the App Store and Google Play mobile apps with the goal of recommending an (English-speaking, free) app profile that can be profitable for both markets.

Based on information from both the App Store and Google Play datasets, a practical app could be most likely to capture users' interests and needs, thus warranting a higher number of installs and providing more opportunities for users to view in-app ads. We also want an app with which users will spend more time using, further maximizing revenue from in-app ads viewed.

With this, I have suggested two ad profiles:

1. An intersection between the reference and travel categories, a directory app for a certain type of business or public service shows promise. Public restrooms or dog parks are two ideas. The directory can have enhanced features like maps, directions, suggestions, and user ratings and reviews.

2. Judging by the popularity of some book apps like for the Bible and the Quran, as well as their supplementary apps, we can create an app for another popular, and ideally more recent book. Beyond the raw text, readers would be able interact with the text (highlight or add notes) as well as access additional features like illustrations, quizzes, and reader fora to enrich their experience and understanding.